core/services/ocr2/plugins/ocr2keeper/evm21/registry.go

```diff
@@ -926,3 +937,75 @@ func (r *EvmRegistry) getBlockHash(blockNumber *big.Int) (common.Hash, error) {
 
 	return block.Hash(), nil
 }
+
+func (r *EvmRegistry) getTxBlock(txHash common.Hash) (*big.Int, common.Hash, error) {
+	txr, err := r.client.TransactionReceipt(r.ctx, txHash)
+	if err != nil {
+		return nil, common.Hash{}, err
+	}
+
+	return txr.BlockNumber, txr.BlockHash, nil
+}
+
+func (r *EvmRegistry) getCheckResult(p ocr2keepers.UpkeepPayload, failureReason uint8) ocr2keepers.CheckResult {
+	return ocr2keepers.CheckResult{
+		FailureReason: failureReason,
+		UpkeepID:      p.UpkeepID,
+		Trigger:       p.Trigger,
+		WorkID:        p.WorkID,
+		FastGasWei:    big.NewInt(0),
+		LinkNative:    big.NewInt(0),
+	}
+}
+
+// verifyCheckBlock checks that the check block and hash are valid
+func (r *EvmRegistry) verifyCheckBlock(checkBlock, upkeepId *big.Int, checkHash common.Hash, p ocr2keepers.UpkeepPayload, i int, results []ocr2keepers.CheckResult) bool {
+	// verify check block number is not too old
+	if r.bs.latestBlock-checkBlock.Int64() > validCheckBlockRange {
```

Comment:
accessing `r.bs.latestBlock` like that can lead to a race..

using `atomic.Int64` for `latestBlock` sounds like the best solution,
but you can also wrap it with a function, e.g. `getLatestBlock` and use a read lock to protect.

Reply:
done